### ISSUE TITLE
Thorlabs Homing Issue

### ISSUE THREAD

**Issue by mattgoode0** [AlexShkarin/pyLabLib#8]
Howdy!

I am working some control software for a ThorLabs K10CR1. I noticed that if the unit has been previously homed through the ThorLabs Kinesis software, the home() function works perfectly. However, if the unit has not been homed through the Kinesis software, the unit is unable to home. The software pauses at this line of code and does not complete. Other functions such as move_to(), is_homed(), and other functions still work. Any ideas on what may be causing this? 

Thanks!

**Comment by AlexShkarin** [comment]
Hi!

Unfortunately, I haven't used this code with K10CR1 (it was contributed by someone else, I only worked with KDC101, which has similar interface), so my ability to help is somewhat limited. I can suggest a couple of things to check for starters:

1) You say that `.is_homed()` works. Does it mean that it also returns the correct status? I.e., would it return `False` on device restart (when it's not homed), and `True` after you home it in the Kinesis software?
2) Does the `.home()` method case the device do anything at all? That is, does the motor start moving when you call it?

Basically, the homing function first sends the homing request to the device, and then just waits in a loop up until the status indicates that it's homed (by the way, you can also make it not wait by calling `.home(sync=False)`, in which case it returns immediately). So, I'd start by figuring out which of the two parts fails: the request (so it doesn't move at all), or the status (so it waits for the status change which never happens).

**Comment by mattgoode0** [comment]
Thanks for the reply! 

1. .is_home() works as intended. False is returned upon device restart and True is returned if the item has been homed through the Kinesis software. 
2. When .home() is commanded, the active LED on the device turns on but the motor does not start moving. 

Please let me know if there is any other information that I can supply! Your help is greatly appreciated.

**Comment by AlexShkarin** [comment]
Ok, so it looks like it's the homing command after all.

Unfortunately, I don't have much to recommend at this point. I know that the command works in principle (I tested it on KDC101 motor), and according to the manual it should also work on K10CR1, but I'm not sure if it works in practice, or what can interfere with it. The only hypothesis I have right now is that homing or limiting settings are somehow off (e.g., homing speed is very slow). To check that, I'd recommend the following:

1. You can get homing parameters (in their raw form) by executing
```python
dev.send_comm_nodata(0x0441)
home_par=dev.recv_comm_data()
```
The result `home_par` should be a `CommData` object with `data` field containing bytestring with the settings. To get their meaning, you can either consult APT manual (https://www.thorlabs.com/Software/Motion%20Control/APT_Communications_Protocol.pdf) page 71, or post it here for me to take a look. In the latter case, you should also post homing parameters (direction, velocity, limit switch) from the Kinesis/APT software.

2. It might also be worthwhile to check limit switch parameters. For that, do the same as above, only with `0x0424` instead of `0x0441`. The result is described on page 73 of the manual.

3. One issue discussion in another software (https://github.com/qpit/thorlabs_apt/issues/3) suggest that re-applying homing parameters in your script might help. To do this, you can run
```python
dev.send_comm_nodata(0x0441)
home_par=dev.recv_comm_data()
dev.send_comm_data(0x0440,home_par.data)
```

4. If that doesn't help, that you might want to consider to factory-reset the device (make sure to record all the relevant settings somewhere).

**Comment by mattgoode0** [comment]
Thanks for the help! 

When I ran the device, I noticed that the unit outputs two different homing parameters based upon if the device has been homed through the Kinesis software or not. If you plug in the device and run 
`dev.send_comm_nodata(0x0441)
home_par=dev.recv_comm_data()`
the following is outputted:
CommData(messageID=1090, data=b'\x01\x00\x02\x00\x01\x00K\x1e\x01\x00\x00\x80\x0c\x00', source=80, dest=1)

If you run the homing process through Kinesis then rerun the code listed above, the following is outputted:
CommData(messageID=1090, data=b"\x01\x00\x02\x00\x01\x00'{^\x04UU\x08\x00", source=80, dest=1)

To resolve the issue I was facing, I ended up sending the second CommData output to the stage upon startup. If I then run `.home()`  everything works perfectly and it does not matter if the stage has been homed through the Kinesis software. 

I don't know if this would be helpful or not, but my default homing parameters are as follows:

- Direction: (reverse)
- Limit Switch: (Reverse (Hard))
- Zero Offset: 4 degrees
- Velocity: 10 degrees per second

Once again, thank you for all the help!

**Comment by AlexShkarin** [comment]
Thank you for the report! I'm glad it worked out!

It looks like in the end the problem was in the homing velocity. The APT calibration is 73300775 microsteps/sec, which amounts to 10 deg/sec; the built-in calibration is only 73291, or 0.01 deg/sec. So, I guess, it was homing before, just extremely slowly. In addition, the zero offset (i.e., the distance from the limit switch which is set to mean zero) is different (4 deg vs. 6 deg), but this is, probably, less important.

I'm not sure how this could have happened. Perhaps, it was shipped with these weird settings, or they were changed some time in the past (although I'm not sure if these startup parameters can even be changed by a user). But, in any case, thank you for letting me know what was the problem!
